File: 04-projects/recipe-assistant/assistant-deno/recipe_insights.ts
// recipe_insights.ts

/**
 * This module provides functionality for generating insights from recipe data using OpenAI's language model.
 * 
 * It directly interacts with OpenAI's API to process recipe information.
 * The module includes logging capabilities for tracking the execution flow and debugging purposes.
 */

import OpenAI from "https://deno.land/x/openai@v4.20.1/mod.ts";
import * as log from "https://deno.land/std@0.192.0/log/mod.ts";

// Configure logging
await log.setup({
  handlers: {
    console: new log.handlers.ConsoleHandler("INFO"),
    file: new log.handlers.FileHandler("INFO", {
      filename: "./app.log",
      formatter: "{levelName} {datetime} {msg}",
    }),
  },
  loggers: {
    default: {
      level: "INFO",
      handlers: ["console", "file"],
    },
  },
});

const client = new OpenAI();

/**
 * Generate insights from a given set of recipes using OpenAI's language model.
 * 
 * This function takes a summary of recipe data as input and uses OpenAI's GPT model to generate
 * insights about trends in cuisine types, popular ingredients, and cooking methods. It aims to
 * provide exactly 5 concise insights, formatted as 'Insight X: Your insight here'.
 * 
 * @param recipes A string containing a summary of recipe data.
 * @returns A dictionary containing at least 3 insights. Keys are formatted as 'insight_1',
 *          'insight_2', etc., and values are the corresponding insight texts.
 * 
 * Note:
 *     If the OpenAI model generates fewer than 3 insights, the function will pad the result
 *     with placeholder insights to ensure at least 3 are returned.
 */
export async function generateInsights(recipes: string): Promise<Record<string, string>> {
  log.info("Starting generateInsights function");
  log.info(`Input recipes: ${recipes}`);

  const prompt = `Based on the following recipe data summary, please provide exactly 5 concise insights about trends in cuisine types, popular ingredients, and cooking methods. Format each insight as 'Insight X: Your insight here', where X is the number of the insight.\n\n${recipes}`;

  log.info("Sending request to OpenAI");
  const chatCompletion = await client.chat.completions.create({
    messages: [{ role: "user", content: prompt }],
    model: "gpt-4-1106-preview",
  });

  const response = chatCompletion.choices[0].message.content;
  log.info(`Generated insights from OpenAI: ${response}`);

  // Parse the response into a dictionary
  const insights: Record<string, string> = {};
  if (response) {
    response.split("\n").forEach((line) => {
      if (line.startsWith("Insight")) {
        const [key, value] = line.split(":", 2);
        insights[key.trim().toLowerCase().replace(" ", "_")] = value.trim();
      }
    });
  }

  // Ensure we have at least 3 insights
  for (let i = 1; Object.keys(insights).length < 3; i++) {
    insights[`insight_${i}`] = "No insight provided";
  }

  log.info(`Parsed insights: ${JSON.stringify(insights)}`);
  log.info("Finished generateInsights function");

  return insights;
}

File: 04-projects/recipe-assistant/assistant-go/recipe_insights.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	"github.com/sashabaranov/go-openai"
)

// logger is a package-level variable for logging
var logger *log.Logger

// init initializes the logger
func init() {
	// Configure logging to write to a file
	file, err := os.OpenFile("app.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatal(err)
	}

	logger = log.New(file, "", log.LstdFlags)
}

// GenerateInsights takes a string of recipe data and returns a map of insights
// It orchestrates the entire process of generating insights
func GenerateInsights(recipes string) (map[string]string, error) {
	start := time.Now()
	logger.Println("Starting GenerateInsights function")
	
	// Create OpenAI client
	client := createOpenAIClient()
	
	// Create prompt for OpenAI
	prompt := createPrompt(recipes)
	
	// Get insights from OpenAI
	insights, err := getInsightsFromOpenAI(client, prompt)
	if err != nil {
		return nil, err
	}
	
	// Parse the insights into a map
	parsedInsights := parseInsights(insights)
	
	// Ensure we have at least 3 insights
	ensureMinimumInsights(parsedInsights, 3)
	
	// Log the results and execution time
	logResults(parsedInsights, start)
	
	return parsedInsights, nil
}

// createOpenAIClient creates and returns an OpenAI client
func createOpenAIClient() *openai.Client {
	clientStart := time.Now()
	client := openai.NewClient(os.Getenv("OPENAI_API_KEY"))
	logger.Printf("Client creation took: %v", time.Since(clientStart))
	return client
}

// createPrompt generates the prompt for OpenAI based on the recipe data
func createPrompt(recipes string) string {
	promptStart := time.Now()
	prompt := fmt.Sprintf("Based on the following recipe data summary, please provide exactly 5 concise insights about trends in cuisine types, popular ingredients, and cooking methods. Format each insight as 'Insight X: Your insight here', where X is the number of the insight.\n\n%s", recipes)
	logger.Printf("Prompt creation took: %v", time.Since(promptStart))
	return prompt
}

// getInsightsFromOpenAI sends a request to OpenAI and returns the response
func getInsightsFromOpenAI(client *openai.Client, prompt string) (string, error) {
	logger.Println("Sending request to OpenAI")
	apiCallStart := time.Now()
	
	// Create a chat completion request
	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: openai.GPT4TurboPreview,
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleUser,
					Content: prompt,
				},
			},
		},
	)
	apiCallDuration := time.Since(apiCallStart)
	logger.Printf("OpenAI API call took %v", apiCallDuration)

	if err != nil {
		logger.Printf("Error from OpenAI: %v", err)
		return "", err
	}

	response := resp.Choices[0].Message.Content
	logger.Printf("Generated insights from OpenAI: %s", response)
	return response, nil
}

// parseInsights converts the OpenAI response into a map of insights
func parseInsights(response string) map[string]string {
	insights := make(map[string]string)
	for _, line := range strings.Split(response, "\n") {
		if strings.HasPrefix(line, "Insight") {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) == 2 {
				key := strings.ToLower(strings.ReplaceAll(strings.TrimSpace(parts[0]), " ", "_"))
				insights[key] = strings.TrimSpace(parts[1])
			}
		}
	}
	return insights
}

// ensureMinimumInsights makes sure we have at least the specified number of insights
func ensureMinimumInsights(insights map[string]string, minimum int) {
	for i := 1; len(insights) < minimum; i++ {
		insights[fmt.Sprintf("insight_%d", i)] = "No insight provided"
	}
}

// logResults logs the final insights and execution time
func logResults(insights map[string]string, start time.Time) {
	insightsJSON, _ := json.Marshal(insights)
	logger.Printf("Parsed insights: %s", string(insightsJSON))
	logger.Println("Finished GenerateInsights function")
	logger.Printf("GenerateInsights function took %v", time.Since(start))
}

func main() {
	// Main function can be implemented here if needed
}